core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -204,6 +187,58 @@ Map<TopicIdPartition, FetchRequest.PartitionData> acquirablePartitions() {
+     */
+    Map<TopicIdPartition, FetchPartitionData> replicaManagerFetchData(Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData,
+                                                                      boolean hasRequestTimedOut) {
+        log.trace("Fetchable share partitions data: {} with groupId: {} fetch params: {}", topicPartitionData,
+            shareFetchData.groupId(), shareFetchData.fetchParams());
+        Map<TopicIdPartition, FetchPartitionData> replicaManagerFetchSatisfyingMinBytes = new HashMap<>();
+        Map<TopicIdPartition, FetchPartitionData> responseData = new HashMap<>();
+        try {
+            Seq<Tuple2<TopicIdPartition, LogReadResult>> responseLogResult = replicaManager.readFromLog(
+                shareFetchData.fetchParams(),
+                CollectionConverters.asScala(
+                    topicPartitionData.entrySet().stream().map(entry ->
+                        new Tuple2<>(entry.getKey(), entry.getValue())).collect(Collectors.toList())
+                ),
+                QuotaFactory.UnboundedQuota$.MODULE$,
+                true);
+
+            AtomicInteger accumulatedBytes = new AtomicInteger(0);
+
+            responseLogResult.foreach(tpLogResult -> {
+                TopicIdPartition topicIdPartition = tpLogResult._1();
+                LogReadResult logResult = tpLogResult._2();
+                FetchPartitionData fetchPartitionData = logResult.toFetchPartitionData(false);
+                responseData.put(topicIdPartition, fetchPartitionData);
+                accumulatedBytes.addAndGet(logResult.info().records.sizeInBytes());
+                return BoxedUnit.UNIT;
+            });
+            log.trace("Data successfully retrieved by replica manager: {}", responseData);
+
+            if (accumulatedBytes.get() >= shareFetchData.fetchParams().minBytes)
+                replicaManagerFetchSatisfyingMinBytes = responseData;
+        } catch (Exception e) {
+            log.error("Error processing delayed share fetch request", e);
+        } finally {
+            // The case when we cannot satisfy the share fetch requests because the response has lesser data than minBytes
+            // and the call is coming from tryComplete, hence we want to release partitions lock so that the next
+            // tryComplete/onComplete call complete successfully.
+            if (replicaManagerFetchSatisfyingMinBytes.isEmpty() && !hasRequestTimedOut) {
+                // Releasing the lock to move ahead with the next request in queue.
+                releasePartitionLocks(shareFetchData.groupId(), topicPartitionData.keySet());
+            }
+        }
+        // This call is coming from onComplete, hence we return the response data irrespective of whether minBytes is
+        // satisfied or not.
+        if (hasRequestTimedOut)
+            return responseData;
```

Comment:
Hmmm, is this same for regular fetch operations as well?

Reply:
Hi @apoorvmittal10, I went through the code for onComplete of `DelayedFetch` ([code reference](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/DelayedFetch.scala#L176)), it seems that it returns whatever it gets from `replicaManager.readFromLog`, so the current code is correct in that case. Please let me know if I am wrong. cc - @junrao